packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD1.hpp

```diff
@@ -0,0 +1,985 @@
+
+      //set the initial coloring of the kh.get_graph_coloring_handle() to be
+      //the data view from the femv.
+      auto femvColors = femv->template getLocalView<MemorySpace>();
+      auto  sv = subview(femvColors, Kokkos::ALL, 0);
+      kh.get_graph_coloring_handle()->set_vertex_colors(sv);
+      kh.get_graph_coloring_handle()->set_tictoc(verbose);
+      KokkosGraph::Experimental::graph_color_symbolic<KernelHandle, lno_row_view_t, lno_nnz_view_t>
+                                                     (&kh, nVtx, nVtx, offset_view, adjs_view);
+      numColors = kh.get_graph_coloring_handle()->get_num_colors();
+
+      if(verbose){
+        std::cout<<"\nKokkosKernels Coloring: "<<kh.get_graph_coloring_handle()->get_overall_coloring_time()<<" iterations: "<<kh.get_graph_coloring_handle()->get_num_phases()<<"\n\n";
+      }
+    }
+    
+    template <class ExecutionSpace, typename MemorySpace>
+    void detectConflicts(const size_t n_local, const size_t n_ghosts,
+		         Kokkos::View<offset_t*, Kokkos::Device<ExecutionSpace, MemorySpace>> dist_offsets,
+			 Kokkos::View<lno_t*, Kokkos::Device<ExecutionSpace, MemorySpace>> dist_adjs,
+			 Kokkos::View<int*, Kokkos::Device<ExecutionSpace, MemorySpace>> femv_colors,
+			 Kokkos::View<lno_t*,
+			              Kokkos::Device<ExecutionSpace, MemorySpace>,
+				      Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_send_atomic,
+		         Kokkos::View<size_t[1],
+			              Kokkos::Device<ExecutionSpace, MemorySpace>,
+				      Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_send_size_atomic,
+			 Kokkos::View<gno_t[1], Kokkos::Device<ExecutionSpace, MemorySpace>> recoloringSize,
+			 Kokkos::View<int*,
+			              Kokkos::Device<ExecutionSpace, MemorySpace>> rand,
+			 Kokkos::View<gno_t*,
+			              Kokkos::Device<ExecutionSpace, MemorySpace>> gid,
+                         Kokkos::View<gno_t*,
+			              Kokkos::Device<ExecutionSpace, MemorySpace>> ghost_degrees,
+			 bool recolor_degrees){
+      Kokkos::parallel_reduce("Conflict Detection",
+		              Kokkos::RangePolicy<ExecutionSpace>(0,n_ghosts), 
+		              KOKKOS_LAMBDA(const int& i,
+				            gno_t& recoloring_size){
+        lno_t lid = i+n_local;
+        int currColor = femv_colors(lid);
+	int currDegree = ghost_degrees(i);
+        for(offset_t j = dist_offsets(lid); j < dist_offsets(lid+1); j++){
+          int nborColor = femv_colors(dist_adjs(j));
+	  int nborDegree = 0;
+	  if((size_t)dist_adjs(j) < n_local) nborDegree = dist_offsets(dist_adjs(j)+1) - dist_offsets(dist_adjs(j));
+	  else nborDegree = ghost_degrees(dist_adjs(j) - n_local); 
+          if(currColor == nborColor ){
+	    if(currDegree < nborDegree && recolor_degrees){
+	      femv_colors(lid) = 0;
+	      recoloring_size++;
+	    }else if (nborDegree < currDegree && recolor_degrees){
+	      femv_colors(dist_adjs(j)) = 0;
+	      recoloring_size++;
+            }else if(rand(lid) > rand(dist_adjs(j))){
```

Comment:
Not sure this pseudo-random tie breaking works correctly. They key is to ensure neighboring processes resolve conflicts the same way. To ensure global consistency, I think we need to use global ids not local ids. Local ids will differ on different processes.

Reply:
Actually, this might work since rand is a vector corresponding to vertices not a function (but it was a bit confusing to me).

Reply:
This only works because I am passing around a rand vector as you noted in a comment below this one. Would the rand generator work in a device kernel? I agree that on-the-fly generation would be less cumbersome than passing a rand view around.